File: plugins/plugin-cron/rust/src/actions/update_cron.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::schedule::{format_schedule, parse_natural_language_schedule};
use crate::types::JobUpdate;
use crate::{Action, ActionExample, ActionResult, CronService};

pub struct UpdateCronAction;

#[async_trait]
impl Action for UpdateCronAction {
    fn name(&self) -> &str {
        "UPDATE_CRON"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "MODIFY_CRON",
            "EDIT_CRON",
            "CHANGE_CRON",
            "ENABLE_CRON",
            "DISABLE_CRON",
            "PAUSE_CRON",
            "RESUME_CRON",
        ]
    }

    fn description(&self) -> &str {
        "Updates an existing cron job. Can pause/resume, change schedules, or modify other properties."
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        let has_update = text.contains("update")
            || text.contains("modify")
            || text.contains("edit")
            || text.contains("change")
            || text.contains("enable")
            || text.contains("disable")
            || text.contains("pause")
            || text.contains("resume");

        let has_cron = text.contains("cron") || text.contains("job") || text.contains("schedule");

        has_update && has_cron
    }

    async fn handler(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&mut CronService>,
    ) -> ActionResult {
        let service = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Cron service is not available.".to_string(),
                    data: None,
                    error: Some("missing_service".to_string()),
                }
            }
        };

        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        // Extract job identifier
        let job_id = extract_job_id(text, service);

        let job_id = match job_id {
            Some(id) => id,
            None => {
                return ActionResult {
                    success: false,
                    text: "Please specify which cron job to update (by ID or name).".to_string(),
                    data: None,
                    error: Some("No job identifier".to_string()),
                }
            }
        };

        // Parse update intent
        let updates = parse_update_intent(text);

        if updates.name.is_none()
            && updates.schedule.is_none()
            && updates.state.is_none()
            && updates.payload.is_none()
        {
            return ActionResult {
                success: false,
                text: "Please specify what to update (e.g. pause, resume, change schedule)."
                    .to_string(),
                data: None,
                error: Some("No updates specified".to_string()),
            };
        }

        match service.update_job(&job_id, updates) {
            Ok(job) => {
                let schedule_str = format_schedule(&job.schedule);
                ActionResult {
                    success: true,
                    text: format!(
                        "Updated cron job \"{}\" ({})\n- Schedule: {}\n- State: {:?}",
                        job.name, job.id, schedule_str, job.state
                    ),
                    data: serde_json::to_value(&job).ok(),
                    error: None,
                }
            }
            Err(e) => ActionResult {
                success: false,
                text: format!("Failed to update job: {}", e),
                data: None,
                error: Some(e),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                user_message: "Pause the cron job called daily-check".to_string(),
                agent_response: "Updated cron job \"daily-check\" (abc-123)\n- State: Paused"
                    .to_string(),
            },
            ActionExample {
                user_message: "Change cron abc-123 to run every 2 hours".to_string(),
                agent_response:
                    "Updated cron job \"status checker\" (abc-123)\n- Schedule: every 2 hours"
                        .to_string(),
            },
        ]
    }
}

/// Extracts a job ID from text (UUID or name lookup).
fn extract_job_id(text: &str, service: &CronService) -> Option<String> {
    // Try UUID pattern
    let uuid_re = regex::Regex::new(
        r"(?i)([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})",
    )
    .ok()?;
    if let Some(caps) = uuid_re.captures(text) {
        return Some(caps.get(1)?.as_str().to_string());
    }

    // Try quoted name
    let quoted_re = regex::Regex::new(r#"["']([^"']+)["']"#).ok()?;
    if let Some(caps) = quoted_re.captures(text) {
        let name = caps.get(1)?.as_str();
        if let Some(job) = service.find_job_by_name(name) {
            return Some(job.id.clone());
        }
    }

    // Try "called/named X"
    let named_re = regex::Regex::new(r"(?i)(?:called|named)\s+(\S+)").ok()?;
    if let Some(caps) = named_re.captures(text) {
        let name = caps.get(1)?.as_str();
        if let Some(job) = service.find_job_by_name(name) {
            return Some(job.id.clone());
        }
    }

    None
}

/// Parses update intent from natural language text.
fn parse_update_intent(text: &str) -> JobUpdate {
    let mut updates = JobUpdate::default();
    let lower = text.to_lowercase();

    // Pause / resume
    if lower.contains("pause") || lower.contains("disable") {
        updates.state = Some(crate::types::JobState::Paused);
    } else if lower.contains("resume") || lower.contains("enable") {
        updates.state = Some(crate::types::JobState::Active);
    }

    // Schedule change
    if let Some(schedule) = parse_natural_language_schedule(text) {
        updates.schedule = Some(schedule);
    }

    // Name change: "rename to X"
    let rename_re = regex::Regex::new(r#"(?i)rename\s+(?:to|as)\s+["']?([^"']+)["']?"#).ok();
    if let Some(re) = rename_re {
        if let Some(caps) = re.captures(text) {
            if let Some(m) = caps.get(1) {
                updates.name = Some(m.as_str().trim().to_string());
            }
        }
    }

    updates
}

File: plugins/plugin-cron/src/actions/update-cron.ts
/**
 * @module update-cron
 * @description Action to update an existing cron job
 */

import type { Action, IAgentRuntime, Memory, State, HandlerCallback } from '@elizaos/core';
import { CronService } from '../services/cron-service.js';
import { CRON_SERVICE_TYPE, CronActions } from '../constants.js';
import type { CronJobPatch, CronSchedule } from '../types.js';
import { formatSchedule, parseScheduleDescription } from '../scheduler/schedule-utils.js';

/**
 * Extracts job identifier (ID or name) from text
 */
function extractJobIdentifier(text: string): { id?: string; name?: string } {
  // Try to find UUID
  const idMatch = /([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})/i.exec(text);
  if (idMatch) {
    return { id: idMatch[1] };
  }

  // Try to find quoted name
  const quotedMatch = /["']([^"']+)["']/i.exec(text);
  if (quotedMatch) {
    return { name: quotedMatch[1] };
  }

  // Try to find "job called X" or "cron named X"
  const namedMatch = /(?:job|cron)\s+(?:called|named)\s+(\S+)/i.exec(text);
  if (namedMatch) {
    return { name: namedMatch[1] };
  }

  return {};
}

/**
 * Parses update intent from text
 */
function parseUpdateIntent(text: string): CronJobPatch {
  const patch: CronJobPatch = {};
  const normalized = text.toLowerCase();

  // Check for enable/disable
  if (normalized.includes('enable') && !normalized.includes('disable')) {
    patch.enabled = true;
  } else if (normalized.includes('disable')) {
    patch.enabled = false;
  }

  // Check for schedule change
  const everyMatch = /every\s+(\d+\s*(?:second|minute|hour|day|week)s?)/i.exec(text);
  if (everyMatch) {
    const schedule = parseScheduleDescription(`every ${everyMatch[1]}`);
    if (schedule) {
      patch.schedule = schedule;
    }
  }

  // Check for name change
  const renameMatch = /rename\s+(?:to|as)\s+["']?([^"']+)["']?/i.exec(text);
  if (renameMatch) {
    patch.name = renameMatch[1].trim();
  }

  return patch;
}

export const updateCronAction: Action = {
  name: CronActions.UPDATE_CRON,
  similes: [
    'MODIFY_CRON',
    'EDIT_CRON',
    'CHANGE_CRON',
    'ENABLE_CRON',
    'DISABLE_CRON',
    'PAUSE_CRON',
    'RESUME_CRON',
  ],
  description:
    'Updates an existing cron job. Can enable/disable jobs, change schedules, or modify other properties.',

  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const text = message.content?.text?.toLowerCase() ?? '';

    const hasUpdateKeyword =
      text.includes('update') ||
      text.includes('modify') ||
      text.includes('edit') ||
      text.includes('change') ||
      text.includes('enable') ||
      text.includes('disable') ||
      text.includes('pause') ||
      text.includes('resume');

    const hasCronKeyword =
      text.includes('cron') || text.includes('job') || text.includes('schedule');

    return hasUpdateKeyword && hasCronKeyword;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: Record<string, unknown>,
    callback?: HandlerCallback
  ) => {
    const cronService = runtime.getService<CronService>(CRON_SERVICE_TYPE);
    if (!cronService) {
      await callback?.({
        text: 'Cron service is not available. Please ensure the plugin is loaded.',
      });
      return { success: false, error: 'Cron service not available' };
    }

    const text = message.content?.text ?? '';

    // Get job ID from options or extract from text
    let jobId: string | undefined = options?.jobId as string | undefined;
    let patch: CronJobPatch = (options?.patch as CronJobPatch) || {};

    if (!jobId) {
      const identifier = extractJobIdentifier(text);

      if (identifier.id) {
        jobId = identifier.id;
      } else if (identifier.name) {
        // Find job by name
        const jobs = await cronService.listJobs({ includeDisabled: true });
        const job = jobs.find(
          (j) => j.name.toLowerCase() === identifier.name!.toLowerCase()
        );

        if (!job) {
          await callback?.({
            text: `No cron job found with name: ${identifier.name}`,
          });
          return { success: false, error: 'Job not found' };
        }

        jobId = job.id;
      }
    }

    if (!jobId) {
      await callback?.({
        text:
          'Please specify which cron job to update. You can use the job ID or name.\n' +
          'Example: "disable cron job abc-123" or "enable cron called daily-check"',
      });
      return { success: false, error: 'No job identifier provided' };
    }

    // If no patch provided in options, parse from text
    if (Object.keys(patch).length === 0) {
      patch = parseUpdateIntent(text);
    }

    if (Object.keys(patch).length === 0) {
      await callback?.({
        text:
          'Please specify what to update. Examples:\n' +
          '- "enable cron job abc-123"\n' +
          '- "disable cron called daily-check"\n' +
          '- "change cron abc-123 to run every 2 hours"',
      });
      return { success: false, error: 'No updates specified' };
    }

    const updatedJob = await cronService.updateJob(jobId, patch);

    const changes: string[] = [];
    if (patch.enabled !== undefined) {
      changes.push(`status: ${patch.enabled ? 'enabled' : 'disabled'}`);
    }
    if (patch.schedule) {
      changes.push(`schedule: ${formatSchedule(patch.schedule)}`);
    }
    if (patch.name) {
      changes.push(`name: ${patch.name}`);
    }

    const nextRun = updatedJob.state.nextRunAtMs
      ? new Date(updatedJob.state.nextRunAtMs).toLocaleString()
      : 'not scheduled';

    await callback?.({
      text:
        `Updated cron job "${updatedJob.name}" (${updatedJob.id})\n` +
        `Changes: ${changes.join(', ')}\n` +
        `Next run: ${nextRun}`,
    });

    return {
      success: true,
      data: {
        jobId: updatedJob.id,
        job: updatedJob,
        changes,
      },
    };
  },

  examples: [
    [
      {
        name: '{{user1}}',
        content: { text: 'Disable the cron job called daily-check' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Updated cron job "daily-check" (abc-123)\nChanges: status: disabled\nNext run: not scheduled',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Enable cron abc-123-def-456' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Updated cron job "status checker" (abc-123-def-456)\nChanges: status: enabled\nNext run: in 5 minutes',
        },
      },
    ],
  ],
};
